src/Language/PureScript/TypeChecker/Entailment/IntCompare.hs

```diff
@@ -0,0 +1,173 @@
+-- |
+-- Graph-based solver for comparing type-level numbers with respect to
+-- reflexivity, symmetry, and transitivity properties.
+--
+module Language.PureScript.TypeChecker.Entailment.IntCompare where
+
+import Protolude
+
+import Control.Arrow ((&&&))
+import qualified Data.Graph as G
+import qualified Data.Map as M
+import qualified Data.Set as S
+
+import qualified Language.PureScript.Types as P
+import qualified Language.PureScript.Constants.Prim as P
+
+data Comparison a
+  = Equal a a
+  | LessThan a a
+  | GreaterThan a a
+  deriving (Show, Eq, Ord)
+
+(.=.) :: a -> a -> Comparison a
+(.=.) = Equal
+
+(.<.) :: a -> a -> Comparison a
+(.<.) = LessThan
+
+(.>.) :: a -> a -> Comparison a
+(.>.) = GreaterThan
+
+type Context a = [Comparison a]
+
+-- | An algorithm for solving comparisons nominally, with respect to
+-- properties like reflexivity, symmetry, and transitivity.
+solveComparison :: forall a. Ord a => Context a -> Comparison a -> Bool
+solveComparison context comparison = case comparison of
+  -- Proving that an equality exists between two nominal values has the
+  -- following requirements:
+  --
+  -- 1. Two types are directly equal.
+  --
+  -- 2. When renamed with respect to all other equalities in the
+  -- context, the renamed comparison must appear in the renamed context.
+  --
+  -- For example, given the context:
+  -- > [ a .=. b, b .=. c, x .=. y, y .=. z ]
+  --
+  -- and when verifying if a comparison holds true:
+  -- > solveComparison context (a .=. c)
+  --
+  -- the context and comparison is renamed as:
+  -- > [ abc .=. abc, xyz .=. xyz ] & (abc .=. abc)
+  Equal a b
+    | a == b -> True
+    | otherwise ->
+        let
+          comparison' = alpha comparison
+          context' = alpha <$> context
+        in
+          comparison' `elem` context'
+  -- Proving that an inequality exists between two nominal values takes
+  -- a very different code path compared to Equal; to summarize, it:
+  --
+  -- 1. Creates a directed graph of all inequalities in the renamed
+  -- context.
+  --
+  -- 2. Searches whether a path exists from point a to point b, allowing
+  -- symmetry to be solved.
+  --
+  -- 3. Also searches whether a path exists from point b to point a,
+  -- solving both the symmetry and reflexivity properties.
+  LessThan a b ->
+    solveInequality fst a b
+  GreaterThan a b ->
+    solveInequality snd a b
+  where
+  solveInequality
+    :: ( ( (G.Graph, S.Set a -> Maybe G.Vertex)
+         , (G.Graph, S.Set a -> Maybe G.Vertex)
+         ) -> (G.Graph, S.Set a -> Maybe G.Vertex)
+       )
+    -> a
+    -> a
+    -> Bool
+  solveInequality f a b =
+    let (graph, search) = f inequalities
+    in fromMaybe False $ do
+      a' <- search $ rename a
+      b' <- search $ rename b
+      let norm = G.path graph a' b'
+          symm = G.path (G.transposeG graph) b' a'
```

Comment:
> Um, isn't symm exactly norm with more steps? There's a path from a' to b' iff there's a path from b' to a' in the graph with all edges reversed?

Yeah, this seems to be left over from my initial algorithm. It passes all tests so this can be removed.

> But also, I think you want to make sure that a' != b'. Consider:

Agreed.